### PR TITLE
disambiguating ambiguous-length float ranges

### PR DESCRIPTION
**Pull request by StefanKarpinski**
This test was occasionally failing (~1 in 10 million times) since there are cases where the length of such a range is ambiguous, i.e. there is more than one `n` such that `start + (n-1)*step == stop`. This adjusts the test to allow any such choice of range length. We should still explicitly disambiguate this case by defining what the length of a range should be in such cases.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Failure of the first commit is unrelated – it's correct and should be retained in the git history. The second commit chooses the length closest to `(stop-start)/step + 1` in ambiguous cases.

File: base/twiceprecision.jl [hunk 1]
```diff
@@ -147,9 +147,16 @@ function colon{T<:Union{Float16,Float32,Float64}}(start::T, step::T, stop::T)
...
-    len += (start < stop′ <= stop) + (start > stop′ >= stop)
+    lf = (stop-start)/step
+    if lf < 0
+        len = 0
+    elseif lf == 0
+        len = 1
+    else
+        len = round(Int, lf) + 1
+        stop′ = start + (len-1)*step
+        len -= (start < stop < stop′) + (start > stop > stop′)
```
Comment (tkelman) [inline base/twiceprecision.jl]: this could use some explanatory comments

Comment (tkelman) [inline base/twiceprecision.jl]: so could the test for that matter

**Comment by tkelman** [comment]
the backport of this is in #20396, correct? (I'll be merging that into tk/backports-0.5.1 shortly)

**Comment by StefanKarpinski** [comment]
Yes, that's it. The situation is slightly different on 0.5 since one of the bugs that came up wasn't introduced until the 0.6 cycle, so I just did the change needed to fix this specific problem along with backporting the tests.

**Pull request merged**